contracts/governance/src/participant.rs

```diff
@@ -0,0 +1,152 @@
+}
+
+impl Participant {
+    pub fn new(address: Address) -> Self {
+        Participant {
+            address,
+            whitelisted: false,
+            current_balance: 0,
+        }
+    }
+
+    pub fn is_whitelisted(&self) -> bool {
+        self.whitelisted
+    }
+
+    pub fn whitelist(&mut self) {
+        self.whitelisted = true
+    }
+
+    pub fn address(&self) -> &Address {
+        &self.address
+    }
+
+    pub fn balance(&self) -> i128 {
+        self.current_balance
+    }
+
+    pub fn increase_balance(&mut self, inc: i128) -> Result<(), Error> {
+        if inc <= 0 {
+            return Err(Error::InvalidAmount);
+        }
+        self.current_balance = self.current_balance.checked_add(inc).unwrap();
+        Ok(())
+    }
+
+    pub fn decrease_balance(&mut self, dec: i128) -> Result<(), Error> {
+        if dec <= 0 {
+            return Err(Error::InvalidAmount);
+        }
+        if dec > self.balance() {
+            return Err(Error::InsufficientFunds);
+        }
+        self.current_balance = self.current_balance.checked_sub(dec).unwrap();
+        Ok(())
+    }
+}
+
+/// Participant repository provides a pattern for
+/// condensing all storage logic for the participant.
+///
+/// It consists on adquiring the participants data structure
+/// from the Soroban storage, do all needed operations and
+/// finally persist them once the object is dropped.
+pub struct Repository<'a> {
+    storage: &'a Persistent,
+    data_key: DataKey,
+    participants_storage: Map<Address, Participant>,
+}
+
+impl<'a> Repository<'a> {
+    pub fn new(storage: &'a Persistent) -> Result<Self, Error> {
+        let participants_storage = storage
+            .get::<_, Map<Address, Participant>>(&DataKey::Participants)
+            .ok_or(Error::AlreadyInitialized)?;
+
+        Ok(Repository {
+            storage,
+            participants_storage,
+            data_key: DataKey::Participants,
+        })
+    }
+
+    pub fn save(&mut self, participant: Participant) {
+        self.participants_storage
+            .set(participant.address.clone(), participant);
+    }
+
+    pub fn find(&mut self, address: Address) -> Result<Participant, Error> {
+        self.participants_storage
+            .get(address)
+            .ok_or(Error::ParticipantNotFound)
+    }
+
+    pub fn remove(&mut self, address: Address) -> Result<(), Error> {
+        self.participants_storage
+            .remove(address)
+            .ok_or(Error::ParticipantNotFound)
+    }
+
+    pub fn whitelisted_balance(&self, env: &Env) -> Map<Address, i128> {
+        self.participants_storage.iter().fold(
+            Map::<Address, i128>::new(env),
+            |mut whitelisted_balance, (addr, participant)| {
+                if participant.is_whitelisted() {
+                    whitelisted_balance.set(addr, participant.balance())
+                }
+                whitelisted_balance
+            },
+        )
+    }
+}
+
+impl<'a> Drop for Repository<'a> {
```

Comment:
The problem i tried to solve here is that i was always missing to do:

```rust
self.storage.set(&self.data_key, &self.participants_storage)
```

Which is updating the storage key with the participant DS. I thought also in leaving a `close()` method (explicit), but then we would be in same problem 🤷 . So putting all of this in a balance, i strived for making it implicit and provide a better experience to the user. i could be wrong thought.

Reply:
The problem is that storage is not free.

Reply:
I dont get it. Wouldnt this be the same like doing it explicit ? :thinking:

Reply:
The problem is drops are implicit. Anyone who creates an instance whether knowingly or unknowingly will have storage.set called.
My point is don't do it on Drop, Drop is for cleaning up memory etc.

Reply:
I am not sure what is not clear here.
Operations have fees. https://stellar.org/blog/developers/sorobans-fee-structure-contributes-stellar-network-scalability

Reply:
I think we are having different discussions on this thread :) . Lets split them.

* Regarding the cost of the storage. Initially i do not consider this a reason to worry about, since calling a method like `close()` for finally commiting the data structure into the persistent storage, would be the same as its currently in the Drop trait impl, which is the execution of `self.storage.set(&self.data_key, &self.participants_storage)`.

* The uses of `Drop`. Under my POV, the use of drop its just attached to that lifecycle part of the instance. We could be releasing memory, closing a socket, [returning a DB connection to a pool](https://github.com/launchbadge/sqlx/blob/ca608a964ec547cb9137e26a29c048f839aa0391/sqlx-core/src/pool/connection.rs#L165-L173)  ... In our case, we are putting back, after doing in memory changes, the participants storage DS on its host storage slot.

* Using multiples instances of the repository. I would not recommend this and i agree it should be explicit in the type documentation. A unique instance its expected in runtime, as 2 separate instances would be operating the same storage key, causing overwrites. So  creating the instance just before the use of it and borrow it in subsequent uses is the way to go. I can make this explicit in docs.

The main motivation of doing this is:
* To avoid an users forgetfulness when persisting data.
* Encapsulate the storage logic, improving code readability.

I currently cannot find strong reasons for not using `Drop` for this. Let me know your opinion on each point and if i am missing any detail :)

Reply:
I also think it's better to make it explicit and maybe do a sync with the storage after each operation. That way we have full control over it + it is testable.

Reply:
In general i think `Drop` is deterministic if everything goes well during execution of that stack. Well it should :sweat_smile:  , if not , none of the previously provided examples would rely on it.

> rust-lang/rust#58874

I cannot find how that would negatively play with this case. For us, `Drop` only needs to be called if everything goes well. In case of a panic, the behaviour explained in such issue, even favours our case, as `Drop` wouldnt be called on a panic, thus saving a call to a host function when [nothing is going to be persisted anyways](https://soroban.stellar.org/docs/fundamentals-and-concepts/contract-lifecycle#execution).

> Each contract execution continues until the contract either completes successfully or traps with an error condition. If execution completes successfully, all ledger entries modified during execution will be written back to the ledger atomically. If execution traps, all modified ledger entries will be discarded and the contract will have no effect on the ledger.

Regarding testing storage related code, AFAIK, we cannot execute storage related code in tests. Only when we are inside a contract.

```rust
#[test]
fn storage_test() {
    Env::default().storage().persistent().set(&1, &1);
}
````

```bash
   1: [Diagnostic Event] topics:[error, Error(Context, MissingValue)], data:"no contract running"
```

Regarding having more control over it. That could be a fair point, but i would wait to an scenario that clearly claims such need of control. We are always in time to change it :)

Let me know your thoughts please. I am open to change it, but i would like to better understand what are the motivations behind.